python/ray/experimental/workflow/workflow_storage.py

```diff
@@ -0,0 +1,268 @@
+"""
+This module is higher-level abstraction of storage directly used by
+workflows.
+"""
+
+import json
+import pathlib
+from typing import Dict, List, Optional, Any, Callable, Tuple, Union
+
+from dataclasses import dataclass
+
+import ray
+from ray.experimental.workflow import storage
+from ray.experimental.workflow.common import Workflow, WorkflowInputs, StepID
+from ray.experimental.workflow import workflow_context
+from ray.experimental.workflow import serialization_context
+
+from ray.experimental.workflow.constants import (
+    STEPS_DIR, OBJECTS_DIR, STEP_INPUTS_METADATA, STEP_OUTPUTS_METADATA,
+    STEP_FUNC_BODY, STEP_ARGS, STEP_OUTPUT, STEP_OUTPUTS_FORWARD)
+
+
+@dataclass
+class StepInspectResult:
+    args_valid: bool = False
+    func_body_valid: bool = False
+    object_refs: Optional[List[str]] = None
+    workflows: Optional[List[str]] = None
+    output_object_valid: bool = False
+    output_step_id: Optional[StepID] = None
+
+    def is_recoverable(self) -> bool:
+        return (self.args_valid and self.object_refs is not None
+                and self.workflows is not None and self.func_body_valid)
+
+
+class WorkflowStorage:
+    """Access workflow in storage. This is a high-level function,
+    which does not care about the underlining storage implementation."""
+
+    def __init__(self,
+                 workflow_id: Optional[str] = None,
+                 store: Optional[storage.Storage] = None):
+        if workflow_id is None:
+            context = workflow_context.get_workflow_step_context()
+            workflow_id = context.workflow_id
+        if store is None:
+            store = storage.get_global_storage()
+        self._storage = store
+        self._workflow_dir = pathlib.Path(workflow_id)
+        self._steps_dir = self._workflow_dir / STEPS_DIR
+        self._objects_dir = self._workflow_dir / OBJECTS_DIR
```

Comment:
Can we remove the use of Path and JSON from this class (move it into storage?) It's weird that the implementation detail of serialization and storage is present in both Storage and WorkflowStorage.